File: Counting_Sort/Counting_Sort.java
package Counting_Sort;

import java.util.Arrays;

public class Counting_Sort {
    /**
     * Sorts an array of integers using the Counting Sort algorithm.
     * 
     * @param arr - The array to be sorted.
     * @return A sorted array.
     */
    public static int[] countingSort(int[] arr) {
        int n = arr.length;
        int[] output = new int[n];

        // Find the maximum value in the array
        int max = arr[0];
        for (int i = 1; i < n; i++) {
            if (arr[i] > max) {
                max = arr[i];
            }
        }

        // Create a counting array to store the occurrences of each element
        int[] count = new int[max + 1];

        // Count the occurrences of each element in the input array
        for (int i = 0; i < n; i++) {
            count[arr[i]]++;
        }

        // Update the counting array to store the cumulative count
        for (int i = 1; i <= max; i++) {
            count[i] += count[i - 1];
        }

        // Build the output array using the counting array
        for (int i = n - 1; i >= 0; i--) {
            output[count[arr[i]] - 1] = arr[i];
            count[arr[i]]--;
        }

        return output;
    }

    public static void main(String[] args) {
        // Example usage of Counting Sort
        int[] arr = { 64, 25, 12, 22, 11 };
        System.out.println("Original List: " + Arrays.toString(arr));
        int[] sortedArr = countingSort(arr);
        System.out.println("Sorted List: " + Arrays.toString(sortedArr));
    }
}

File: Counting_Sort/Counting_Sort.cpp
#include <iostream>
#include <algorithm>

using namespace std;

/**
 * Sorts an array of integers using the Counting Sort algorithm.
 * @param arr - The array to be sorted.
 * @param n - The size of the array.
 * @param maxVal - The maximum value present in the array.
 */
void countingSort(int arr[], int n, int maxVal)
{
    // Create an output array and a counting array
    int output[n];
    int count[maxVal + 1];

    // Initialize the counting array with zeros
    for (int i = 0; i <= maxVal; i++)
    {
        count[i] = 0;
    }

    // Count the occurrences of each element in the input array
    for (int i = 0; i < n; i++)
    {
        count[arr[i]]++;
    }

    // Update the counting array to store the cumulative count
    for (int i = 1; i <= maxVal; i++)
    {
        count[i] += count[i - 1];
    }

    // Build the output array using the counting array
    for (int i = n - 1; i >= 0; i--)
    {
        output[count[arr[i]] - 1] = arr[i];
        count[arr[i]]--;
    }

    // Copy the sorted elements back to the original array
    for (int i = 0; i < n; i++)
    {
        arr[i] = output[i];
    }
}

int main()
{
    // Example usage of Counting Sort
    int arr[] = {64, 25, 12, 22, 11};
    int n = sizeof(arr) / sizeof(arr[0]);
    int maxVal = *max_element(arr, arr + n);

    // Print the original array
    cout << "Original List: ";
    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;

    // Call Counting Sort to sort the array
    countingSort(arr, n, maxVal);

    // Print the sorted array
    cout << "Sorted List: ";
    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;

    return 0;
}
